### PR TITLE
Add alphas to control language and speaker balancer

### PR DESCRIPTION
**Pull request by Edresson**
This PR fixes the issue/suggestion reported in #1185 .

It permits the combination of weights for speaker and language balance. I normalize the language and speaker weights and added an independent alpha for each one.I performed the sum of the weights multiplied by the respective alphas of language and speaker balancer. In this way, we can control the influence of speaker and language in the batch balancer and easily add a new balancer if necessary in the future.

### CODE REVIEW

**Comment by vince62s** [comment]
Hi @Edresson . I am not sure to fully understand the logic.
say language_alpha = 10 and speaker_alpha = 1
hos does it behave exactly from a language balancing  standpoint, from a speaker balancing standpoint and from a language/speaker standpoint.
thanks

**Comment by Edresson** [comment]
> Hi @Edresson . I am not sure to fully understand the logic. say language_alpha = 10 and speaker_alpha = 1 hos does it behave exactly from a language balancing standpoint, from a speaker balancing standpoint and from a language/speaker standpoint. thanks

Yeah, It is the idea. We can control the language influence and speaker. In this case, language will generate the highest weights and speaker small ones. During the sample selection in a language X,  a speaker with few samples will have more probably to be chosen than a speaker with more samples, however, we will keep the language balancer intact (because it has a  high alpha). In theory, the alphas can be used to create levels of balancing. Balancers with the highest alphas will have priority (in this case language).

File: TTS/tts/models/base_tts.py [hunk 1]
```diff
@@ -355,12 +356,22 @@ def get_data_loader(
...
+                    print(" > Using Language weighted sampler with alpha:", alpha)
+                    weights = get_language_balancer_weights(dataset.items) * alpha
+
+                if getattr(config, "use_speaker_weighted_sampler", False):
+                    alpha = getattr(config, "speaker_weighted_sampler_alpha", 1.0)
+                    print(" > Using Speaker weighted sampler with alpha:", alpha)
+                    if weights is not None:
+                        weights += get_speaker_balancer_weights(dataset.items) * alpha
+                    else:
+                        weights = get_speaker_balancer_weights(dataset.items) * alpha
```
Comment (erogol) [inline TTS/tts/models/base_tts.py:371]: I think gathering all the samplers and weighters in TTS/tts/datasets/samplers.py and using a factory pattern to pull the right set of samplers would be a more convenient solution.

File: TTS/tts/utils/speakers.py [hunk 2]
```diff
@@ -413,11 +412,13 @@ def get_speaker_manager(c: Coqpit, data: List = None, restore_path: str = None,
...
     unique_speaker_names = np.unique(speaker_names).tolist()
     speaker_ids = [unique_speaker_names.index(l) for l in speaker_names]
     speaker_count = np.array([len(np.where(speaker_names == l)[0]) for l in unique_speaker_names])
     weight_speaker = 1.0 / speaker_count
-    dataset_samples_weight = torch.from_numpy(np.array([weight_speaker[l] for l in speaker_ids])).double()
-    return WeightedRandomSampler(dataset_samples_weight, len(dataset_samples_weight))
+    dataset_samples_weight = np.array([weight_speaker[l] for l in speaker_ids])
+    # normalize
+    dataset_samples_weight = dataset_samples_weight / np.linalg.norm(dataset_samples_weight)
+    return torch.from_numpy(dataset_samples_weight).double()
```
Comment (erogol) [inline TTS/tts/utils/speakers.py:424]: I think ```.float()``` will suffice without bothering the RAM too much ?

File: tests/data_tests/test_samplers.py [hunk 3]
```diff
@@ -4,7 +4,8 @@
```
Comment (erogol) [inline tests/data_tests/test_samplers.py:4]: Can you use https://docs.python.org/3/library/unittest.html in coherence with the rest of the tests?

**Review by erogol — CHANGES_REQUESTED** [review]

**Comment by erogol** [comment]
I like the idea of computing weights and then sampling 👍

File: TTS/tts/models/base_tts.py (hunk 1 above)
Comment (Edresson) [reply TTS/tts/models/base_tts.py:371]: Agreed, but we will change all of it for multi GPU support so I just kept it in the way that it was. Do you want this change or we can do it when we will implement the multi GPU support? (I guess that it will change a lot)

File: TTS/tts/utils/speakers.py (hunk 2 above)
Comment (Edresson) [reply TTS/tts/utils/speakers.py:424]: Indeed, done

File: tests/data_tests/test_samplers.py (hunk 3 above)
Comment (Edresson) [reply tests/data_tests/test_samplers.py:6]: Done

File: TTS/tts/models/base_tts.py (hunk 1 above)
Comment (erogol) [reply TTS/tts/models/base_tts.py:371]: get_data_loader got too long. Somehow we need to split it so that we can override different parts in different models.  

Maybe 

```BaseTTS.get_sampler()``` ?

Comment (Edresson) [reply TTS/tts/models/base_tts.py:371]: Done, I added get_sampler in BaseTTS.

File: TTS/tts/models/base_tts.py [hunk 4]
```diff
@@ -12,11 +12,12 @@
 from TTS.model import BaseModel
 from TTS.tts.configs.shared_configs import CharactersConfig
 from TTS.tts.datasets.dataset import TTSDataset
-from TTS.tts.utils.languages import LanguageManager, get_language_weighted_sampler
-from TTS.tts.utils.speakers import SpeakerManager, get_speaker_manager, get_speaker_weighted_sampler
+from TTS.tts.utils.languages import LanguageManager, get_language_balancer_weights
```
Comment (erogol) [inline TTS/tts/models/base_tts.py:15]: Why don't you compute sampler weights in the Managers? They already have all the info necessary for computing the weights.

File: TTS/tts/models/base_tts.py [hunk 5]
```diff
@@ -234,6 +235,26 @@ def format_batch(self, batch: Dict) -> Dict:
...
+        if getattr(config, "use_speaker_weighted_sampler", False):
+            alpha = getattr(config, "speaker_weighted_sampler_alpha", 1.0)
+            print(" > Using Speaker weighted sampler with alpha:", alpha)
+            if weights is not None:
+                weights += get_speaker_balancer_weights(data_items) * alpha
+            else:
+                weights = get_speaker_balancer_weights(data_items) * alpha
+
+        if weights is not None:
+            sampler = WeightedRandomSampler(weights, len(weights))
```
Comment (erogol) [inline TTS/tts/models/base_tts.py:254]: Is this DP compatible?

**Review by erogol — CHANGES_REQUESTED** [review]

File: TTS/tts/models/base_tts.py (hunk 4 above)
Comment (Edresson) [reply TTS/tts/models/base_tts.py:15]: Indeed, but maybe we can use it in the speaker encoder/vocoder/deep fake detector. In this case, maybe its better to keep it as functions (because in these other models Language/Speaker manager will be not used). But if you like I can move to Language/Speaker manager and in the future, if necessary we can just copy the function.

File: TTS/tts/models/base_tts.py (hunk 5 above)
Comment (Edresson) [reply TTS/tts/models/base_tts.py:254]: Not for now

Comment (Edresson) [reply TTS/tts/models/base_tts.py:256]: Now it is DDP compatible :)

Comment (erogol) [reply TTS/tts/models/base_tts.py:256]: Maybe rename `DistributedSamplerWithSampler` as `DistributedSamplerWrapper` ?

File: TTS/utils/samplers.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
+import torch
```
Comment (erogol) [inline TTS/utils/samplers.py:1]: Maybe loge it in the `trainer.torch`?

that way we can share this among projects if it is generally applicable.

**Review by erogol — CHANGES_REQUESTED** [review]

File: TTS/utils/samplers.py (hunk 6 above)
Comment (Edresson) [reply TTS/utils/samplers.py:1]: Yeah we can do it in a future PR when we merge the  FactorTrainerAPI in dev. Maybe the better is just merging this PR for now.

File: TTS/tts/models/base_tts.py (hunk 5 above)
Comment (Edresson) [reply TTS/tts/models/base_tts.py:256]: Done

File: TTS/config/shared_configs.py [hunk 7]
```diff
@@ -251,11 +251,27 @@ class BaseTrainingConfig(TrainerConfig):
...
+            Enable / Disable the batch balancer by speaker. Defaults to ```False```.
+
+        speaker_weighted_sampler_alpha (float):
+            Number that control the influence of the speaker sampler weights. Defaults to ```1.0```.
+
+        use_language_weighted_sampler (bool):
+            Enable / Disable the batch balancer by language. Defaults to ```False```.
+
+        language_weighted_sampler_alpha (float):
+            Number that control the influence of the language sampler weights. Defaults to ```1.0```.
```
Comment (erogol) [inline TTS/config/shared_configs.py:265]: These arguments are supposed to be used only with TTS models right?

If so why don't you move these to shared TTS configs?

**Review by erogol — CHANGES_REQUESTED** [review]

File: TTS/config/shared_configs.py (hunk 7 above)
Comment (Edresson) [reply TTS/config/shared_configs.py:265]: Indeed, when I implemented it I was planning to use it in the emotion encoder. But it does not make more sense after the implementation of the new sampler.

Comment (Edresson) [reply TTS/config/shared_configs.py]: Done

**Review by erogol — APPROVED** [review]

**Comment by vince62s** [comment]
@Edresson sorry to revive this discussion but the 0.6.2 is about to be merged including this.

as is, let say we have a very abnormal distribution of speakers, eg: a few speakers with very little number of samples. Those will be balanced with a probability which is the same as another speaker with a huge amount of data. Am I correct ?
I think there cold be some edges where some datasets include noisy speakers. I may be wrong though.

**Comment by Edresson** [comment]
> @Edresson sorry to revive this discussion but the 0.6.2 is about to be merged including this.
> 
> as is, let say we have a very abnormal distribution of speakers, eg: a few speakers with very little number of samples. Those will be balanced with a probability which is the same as another speaker with a huge amount of data. Am I correct ? I think there cold be some edges where some datasets include noisy speakers. I may be wrong though.

Yeah, It is the objective of the speaker balancer is that all speakers appear in a similar frequency in the batch. If the user has noisy speakers he needs to remove these speakers from the dataset. Unfortunately, we can't control it :(.

> @Edresson sorry to revive this discussion but the 0.6.2 is about to be merged including this.
> 
> as is, let say we have a very abnormal distribution of speakers, eg: a few speakers with very little number of samples. Those will be balanced with a probability which is the same as another speaker with a huge amount of data. Am I correct ? I think there cold be some edges where some datasets include noisy speakers. I may be wrong though.

Yeah, the objective of the speaker balancer is that all speakers appear in a similar frequency in the batch. If the user has noisy speakers he needs to remove these speakers from the dataset. Unfortunately, we can't control it :(.

**Comment by vince62s** [comment]
Well somehow it is a change in behavior I think vs previous version, so might be good to be specific in the doc.

**Comment by Edresson** [comment]
> Well somehow it is a change in behavior I think vs previous version, so might be good to be specific in the doc.

By default this use_speaker_weighted_sampler and use_language_weighted_sampler are disable. So if the user does not enable it, no change exists between the versions :). The only thing that this PR, changes is that currently, we can enable the language weighted sampler and the Speaker weighted sampler together (and use multiples GPUS with these samplers).
